runtime/cudaq/utils/details/impls/xtensor_impl.cpp

```diff
@@ -0,0 +1,163 @@
+  std::vector<std::size_t> m_shape; ///< Shape of the tensor
+  bool ownsData = true; ///< Flag indicating if this object owns the data
+
+  /// @brief Check if the given indices are valid for this tensor
+  /// @param idxs Vector of indices to check
+  /// @return true if indices are valid, false otherwise
+  bool validIndices(const std::vector<std::size_t> &idxs) const {
+    if (idxs.size() != m_shape.size())
+      return false;
+    for (std::size_t dim = 0; auto idx : idxs)
+      if (idx < 0 || idx >= m_shape[dim++])
+        return false;
+    return true;
+  }
+
+public:
+  /// @brief Constructor for xtensor
+  /// @param d Pointer to the tensor data
+  /// @param s Shape of the tensor
+  xtensor(const Scalar *d, const std::vector<std::size_t> &s)
+      : m_data(const_cast<Scalar *>(d)), m_shape(s) {}
+
+  /// @brief Get the rank of the tensor
+  /// @return The rank (number of dimensions) of the tensor
+  std::size_t rank() const override { return m_shape.size(); }
+
+  /// @brief Get the total size of the tensor
+  /// @return The total number of elements in the tensor
+  std::size_t size() const override {
+    return std::accumulate(m_shape.begin(), m_shape.end(), 1,
+                           std::multiplies<size_t>());
+  }
+
+  /// @brief Get the shape of the tensor
+  /// @return A vector containing the dimensions of the tensor
+  std::vector<std::size_t> shape() const override { return m_shape; }
+
+  /// @brief Access a mutable element of the tensor
+  /// @param indices The indices of the element to access
+  /// @return A reference to the element at the specified indices
+  /// @throws std::runtime_error if indices are invalid
+  Scalar &at(const std::vector<size_t> &indices) override {
+    if (!validIndices(indices))
+      throw std::runtime_error("Invalid tensor indices: " +
+                               fmt::format("{}", fmt::join(indices, ", ")));
+
+    return xt::adapt(m_data, size(), xt::no_ownership(), m_shape)[indices];
+  }
+
+  /// @brief Access a const element of the tensor
+  /// @param indices The indices of the element to access
+  /// @return A const reference to the element at the specified indices
+  /// @throws std::runtime_error if indices are invalid
+  const Scalar &at(const std::vector<size_t> &indices) const override {
+    if (!validIndices(indices))
+      throw std::runtime_error("Invalid constant tensor indices: " +
+                               fmt::format("{}", fmt::join(indices, ", ")));
+    return xt::adapt(m_data, size(), xt::no_ownership(), m_shape)[indices];
+  }
+
+  /// @brief Copy data into the tensor
+  /// @param d Pointer to the source data
+  /// @param shape The shape of the source data
+  void copy(const Scalar *d, const std::vector<std::size_t> &shape) override {
+    auto size = std::accumulate(shape.begin(), shape.end(), 1,
+                                std::multiplies<size_t>());
+    if (m_data)
+      std::free(m_data);
```

Comment:
```suggestion
      delete m_data;
```
This matches the next line, but as noted above the ctor is just assuming ownership of data that was allocated outside of this code.